### PR TITLE
Internalize libraries

### LINKED ISSUE #94: Internalize libraries
IntentSolutionLib, UserIntentLib, AssetBasedIntentCurveLib, RevertReason are libraries that have external functions. While this allows for using ´calldata´ parameters, these libraries come with the baggage of separate deployment.

### PR DESCRIPTION
**Pull request by supiket**
- Changes to `IntentSolutionLib`, `UserIntentLib`, `AssetBasedIntentCurveLib`, `RevertReason`:
  - `public` functions -> `internal` functions
- Additional changes to `AssetBasedIntentCurveLib`, `RevertReason`: 
  - `calldata` parameters -> `memory` parameters
- Introduce `AssetBasedIntentCurveLibHarness` library for testing `AssetBasedIntentCurveLib` functions

Closes #94

### CODE REVIEW

File: src/standards/assetbased/AssetBasedIntentCurve.sol [hunk 1]
```diff
@@ -47,7 +47,7 @@ library AssetBasedIntentCurveLib {
     uint16 private constant FLAGS_CURVE_TYPE_MASK = 0x00fc;
     uint16 private constant FLAGS_ASSET_TYPE_MASK = 0xff00;
 
-    function validate(AssetBasedIntentCurve calldata curve) public pure {
+    function validate(AssetBasedIntentCurve memory curve) internal pure {
```
Comment (pixelcircuits) [inline src/standards/assetbased/AssetBasedIntentCurve.sol:50]: can we pull these functions out of the lib so they can still use calldata?

File: test/AssetBasedIntentCurve.t.sol [hunk 2]
```diff
@@ -4,132 +4,149 @@ pragma solidity ^0.8.13;
...
-    AssetBasedIntentCurve internal _testLinearCurve =
-        _curveETH(AssetBasedIntentCurveBuilder.linearCurve(2, 10, 20, false), EvaluationType.ABSOLUTE);
-    AssetBasedIntentCurve internal _testLinearFlippedCurve =
-        _curveETH(AssetBasedIntentCurveBuilder.linearCurve(2, 10, 20, true), EvaluationType.ABSOLUTE);
-    AssetBasedIntentCurve internal _testExponentialCurve =
-        _curveETH(AssetBasedIntentCurveBuilder.exponentialCurve(2, 10, 2, 20, false), EvaluationType.ABSOLUTE);
-    AssetBasedIntentCurve internal _testExponentialFlippedCurve =
-        _curveETH(AssetBasedIntentCurveBuilder.exponentialCurve(2, 10, 2, 20, true), EvaluationType.ABSOLUTE);
+    function _constantCurve() private pure returns (AssetBasedIntentCurve memory) {
+        return _curveETH(AssetBasedIntentCurveBuilder.constantCurve(10), EvaluationType.ABSOLUTE);
```
Comment (pixelcircuits) [inline test/AssetBasedIntentCurve.t.sol:13]: why did these have to be wrapped into functions?

Comment (supiket) [reply test/AssetBasedIntentCurve.t.sol:14]: not a necessity by any means. looks cleaner and shadows the private functions `_intentForCase` and `_solutionForCase` that are used in scenarios

File: src/test/AssetBasedIntentCurveLibHarness.sol [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+    parseAssetType,
+    parseCurveType,
+    parseEvaluationType,
+    isRelativeEvaluation,
+    CurveType,
+    EvaluationType
+} from "../standards/assetbased/AssetBasedIntentCurve.sol";
+import {AssetType} from "../standards/assetbased/utils/AssetWrapper.sol";
+
+library AssetBasedIntentCurveLibHarness {
```
Comment (supiket) [inline src/test/AssetBasedIntentCurveLibHarness.sol:17]: External functions can receive calldata parameters and `AssetBasedIntentCurveLib` is not an existing external library anymore. For tests, this library is required to test the now free functions of ex-`AssetBasedIntentCurveLib` outside a `handleIntents` call, which itself is an external function and can receive a calldata parameter.

File: test/AssetBasedIntentCurve.t.sol (hunk 2 above)
Comment (pixelcircuits) [reply test/AssetBasedIntentCurve.t.sol:14]: Cool! That's fine

**Review by pixelcircuits — APPROVED** [review]

**Pull request merged**